### PR TITLE
Feature/scaleio volume movement

### PR DESCRIPTION
**Pull request by thanset**

### CODE REVIEW

**Comment by thanset** [comment]
Fixed conflicts.  This adds host filtering for ScaleIO pods in the cluster, as well as volume movement via Rexray/Dvdcli.

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java [hunk 1]
```diff
@@ -416,20 +417,36 @@ private CommandInfo getNewBrokerCmd(KafkaSchedulerConfiguration config, int brok
...
+
+        String brokerName = OfferUtils.brokerIdToTaskName(brokerId);
         BrokerConfiguration brokerConfiguration = config.getBrokerConfiguration();
         ZookeeperConfiguration zookeeperConfiguration = config.getZookeeperConfig();
         ExecutorConfiguration executorConfiguration = config.getExecutorConfiguration();
         String frameworkName = config.getServiceConfiguration().getName();
 
+        // Get rexray option here.
+        StringBuilder stringBuilder = new StringBuilder();
+        if (executorConfiguration.getVolumeDriver().equalsIgnoreCase("rexray")) {
```
Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: Did we want to make this configurable for other stuff like portworx? If not what about making `"rexray"` a `final static String` somewhere?

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java [hunk 2]
```diff
@@ -416,20 +417,36 @@ private CommandInfo getNewBrokerCmd(KafkaSchedulerConfiguration config, int brok
...
+            stringBuilder.append("./dvdcli mount --volumename=");
```
Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: Might be cleaner to have this be one line and use `String.format()`.

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java [hunk 3]
```diff
@@ -416,20 +417,36 @@ private CommandInfo getNewBrokerCmd(KafkaSchedulerConfiguration config, int brok
...
+
         Map<String, String> executorEnvMap = new HashMap<>();
         executorEnvMap.put(JAVA_HOME_KEY, JAVA_HOME_VALUE);
         executorEnvMap.put("FRAMEWORK_NAME", frameworkName);
         executorEnvMap.put("KAFKA_ZOOKEEPER_URI", zookeeperConfiguration.getKafkaZkUri());
         executorEnvMap.put(KafkaEnvConfigUtils.KAFKA_OVERRIDE_PREFIX + "BROKER_ID", Integer.toString(brokerId));
         executorEnvMap.put(CONFIG_ID_KEY, configName);
         return CommandInfo.newBuilder()
-                .setValue(executorConfiguration.getCommand())
+                .setValue(executorCommand)
```
Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: This will break my change of exposing the executor command as a config. We need this so that we can do `ifconfig lo up` until we have this fix https://issues.apache.org/jira/browse/MESOS-6269

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java [hunk 4]
```diff
@@ -416,20 +417,36 @@ private CommandInfo getNewBrokerCmd(KafkaSchedulerConfiguration config, int brok
...
         ExecutorConfiguration executorConfiguration = config.getExecutorConfiguration();
         String frameworkName = config.getServiceConfiguration().getName();
 
+        // Get rexray option here.
+        StringBuilder stringBuilder = new StringBuilder();
+        if (executorConfiguration.getVolumeDriver().equalsIgnoreCase("rexray")) {
+            stringBuilder.append("./dvdcli mount --volumename=");
+            stringBuilder.append(brokerName.replace("broker-", executorConfiguration.getVolumeName() + "_"));
+            stringBuilder.append(" --volumedriver=");
+            stringBuilder.append(executorConfiguration.getVolumeDriver().trim());
```
Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: Is there something that's causing extra white space with this stuff so that you need to `trim()`?

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/scheduler/KafkaScheduler.java [hunk 5]
```diff
@@ -326,21 +329,50 @@ public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
...
+                .collect(Collectors.toList());
+
+    }
+
+    private List<Offer> filterOfferByHostName(List<Offer> offers, List<String> filters){
+        List<Protos.Offer> filteredOffers = new ArrayList<>();
+
+        for (String filter_term : filters) {
+            for (Protos.Offer offer : offers){
+                if (filter_term.trim().equals(offer.getHostname().trim())){
```
Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/scheduler/KafkaScheduler.java]: What causes extra white space and the need for `trim()`s?

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/scheduler/KafkaScheduler.java [hunk 6]
```diff
@@ -326,21 +329,50 @@ public void statusUpdate(SchedulerDriver driver, TaskStatus status) {
...
+                .filter(offer -> offer.getAttributesList().stream().anyMatch(attribute -> attribute.getText()
+                        .equals(Protos.Value.Text.newBuilder().setValue(filter).build())))
+                .collect(Collectors.toList());
+
+    }
+
+    private List<Offer> filterOfferByHostName(List<Offer> offers, List<String> filters){
+        List<Protos.Offer> filteredOffers = new ArrayList<>();
+
+        for (String filter_term : filters) {
```
Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/scheduler/KafkaScheduler.java]: What about having the `filters` be a map so that you only need to iterate over the offers? Not sure if that change is really worth it though, I'm sure the filters list will always be relatively small unless someone deploys with a massive set of their own conditions...

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java (hunk 3 above)
Comment (thanset) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: Since your command is set in the config, it's already there, it won't break your "ifup lo" command. It's just appended with my things as well.

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java (hunk 1 above)
Comment (thanset) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: Good idea to make this configurable.

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/scheduler/KafkaScheduler.java (hunk 6 above)
Comment (thanset) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/scheduler/KafkaScheduler.java]: This is still essentially O(n) since the number of filters will remain a constant during execution and can't be changed at runtime.  If we run into performance problems we can debug and certainly look at this filtering code again.

File: kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java (hunk 2 above)
Comment (thanset) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: http://stackoverflow.com/questions/513600/should-i-use-javas-string-format-if-performance-is-important

tl;dr JIT compiles concat into a string builder anyways if you use +'s.  String format is slower if we care about performance. What do you think? I think style-wise it's easier to read vs .format()

Comment (aaronjwood) [inline kafka-scheduler/src/main/java/com/mesosphere/dcos/kafka/offer/PersistentOfferRequirementProvider.java]: I'm good either way. I thought `.format()` might be easier to read but that might just be my preference. I thought that this method is only run once but if it gets called a lot I think we should definitely go with your suggestion to stay away from `.format()`.
